src/smt/theory_str_noodler/theory_str_noodler_final_check.cpp

```diff
@@ -418,4 +418,35 @@ namespace smt::noodler {
             }
         }
     }
+
+    bool theory_str_noodler::is_nielsen_suitable(const Formula& instance) const {
+        if(this->m_membership_todo_rel.size() != 0 || this->m_not_contains_todo_rel.size() != 0) {
+            return false;
+        }
+        if(!instance.is_quadratic()) {
+            return false;
+        }
+        Graph incl = Graph::create_inclusion_graph(instance);
+        return incl.is_cyclic();
+    }
+
+    bool theory_str_noodler::is_underapprox_suitable(const Formula& instance, const AutAssignment& aut_ass) const {
+        int ln = 0;
+        for(const Predicate& pred : instance.get_predicates()) {
+            for(const BasicTerm& var : pred.get_vars()) {
+                
+                if(aut_ass.at(var)->size() <= 1) {
+                    continue;
+                }
+                if(aut_ass.is_co_finite(var, ln)) {
+                    continue;
+                }
+                if(aut_ass.is_singleton(var)) {
+                    continue;
+                }
```

Comment:
Some reasoning why this condition would be good.

Reply:
I added some comments. Please check it.

Reply:
Looks great. I am happy with the changes.

Reply:
Looks good.